NX 20.2.1, Newly created react-monorepo & module federation setup doesn't work

### Current Behavior

` npx create-nx-workspace react-monorepo --preset=react-monorepo --pm=pnpm`

then choose:
app name = my-app
bundler = rspack
test runner = none
stylesheet = CSS
CI = later
cacheing = no

`cd react-monorepo`

`pnpm nx g @nx/react:host shell --remotes=remote --bundler=rspack`

stylesheet = css
e2e = none

*succeeds*

`nx serve shell`

FAILS with error:

 NX   Remote failed to start. A complete log can be found in: /Volumes/Code/react-monorepo/.nx/workspace-data/2024-12-09T15_30_47_332Z-build.log


LOG

```

 NX   Running target build for project remote:

- remote



> nx run remote:build:production


 NX   Cannot find module '@nx/module-federation/rspack'

Require stack:
- /Volumes/Code/react-monorepo/remote/rspack.config.ts
- /Volumes/Code/react-monorepo/remote/rspack.config.prod.ts
- /Volumes/Code/react-monorepo/node_modules/.pnpm/@nx+rspack@20.2.1_@module-federation+enhanced@0.7.6_@module-federation+node@2.6.11_@swc-node+_vvolop54a2iqvroiphed6tvd2q/node_modules/@nx/rspack/src/utils/resolve-user-defined-rspack-config.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/@nx+rspack@20.2.1_@module-federation+enhanced@0.7.6_@module-federation+node@2.6.11_@swc-node+_vvolop54a2iqvroiphed6tvd2q/node_modules/@nx/rspack/src/executors/rspack/lib/config.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/@nx+rspack@20.2.1_@module-federation+enhanced@0.7.6_@module-federation+node@2.6.11_@swc-node+_vvolop54a2iqvroiphed6tvd2q/node_modules/@nx/rspack/src/utils/create-compiler.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/@nx+rspack@20.2.1_@module-federation+enhanced@0.7.6_@module-federation+node@2.6.11_@swc-node+_vvolop54a2iqvroiphed6tvd2q/node_modules/@nx/rspack/src/executors/rspack/rspack.impl.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/config/schema-utils.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/command-line/run/executor-utils.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/devkit-internals.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/utils/assert-workspace-validity.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/project-graph/build-project-graph.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/project-graph/project-graph.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/project-graph/file-utils.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/utils/package-manager.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/utils/package-json.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/utils/print-help.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/command-line/run/run.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/bin/run-executor.js
Pass --verbose to see the stacktrace.

 NX   Running target build for project remote failed

Failed tasks:

- remote:build:production
```

So it looks like it's just missing  '@nx/module-federation/rspack' package, so I triied to install it, but that didn't work. Realiized I need just @nx/module-federation so installed that, and now it works.

So I think the generators need to be updated to install this package.



### Expected Behavior

Shell app is served with remotes.

### GitHub Repo

_No response_

### Steps to Reproduce


1. ` npx create-nx-workspace react-monorepo --preset=react-monorepo --pm=pnpm`

then choose:
app name = my-app
bundler = rspack
test runner = none
stylesheet = CSS
CI = later
cacheing = no

`cd react-monorepo`

2. `pnpm nx g @nx/react:host shell --remotes=remote --bundler=rspack`

stylesheet = css
e2e = none

*succeeds*

3. `nx serve shell`

FAILS with error:

 NX   Remote failed to start. A complete log can be found in: /Volumes/Code/react-monorepo/.nx/workspace-data/2024-12-09T15_30_47_332Z-build.log


LOG

```

 NX   Running target build for project remote:

- remote



> nx run remote:build:production


 NX   Cannot find module '@nx/module-federation/rspack'

Require stack:
- /Volumes/Code/react-monorepo/remote/rspack.config.ts
- /Volumes/Code/react-monorepo/remote/rspack.config.prod.ts
- /Volumes/Code/react-monorepo/node_modules/.pnpm/@nx+rspack@20.2.1_@module-federation+enhanced@0.7.6_@module-federation+node@2.6.11_@swc-node+_vvolop54a2iqvroiphed6tvd2q/node_modules/@nx/rspack/src/utils/resolve-user-defined-rspack-config.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/@nx+rspack@20.2.1_@module-federation+enhanced@0.7.6_@module-federation+node@2.6.11_@swc-node+_vvolop54a2iqvroiphed6tvd2q/node_modules/@nx/rspack/src/executors/rspack/lib/config.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/@nx+rspack@20.2.1_@module-federation+enhanced@0.7.6_@module-federation+node@2.6.11_@swc-node+_vvolop54a2iqvroiphed6tvd2q/node_modules/@nx/rspack/src/utils/create-compiler.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/@nx+rspack@20.2.1_@module-federation+enhanced@0.7.6_@module-federation+node@2.6.11_@swc-node+_vvolop54a2iqvroiphed6tvd2q/node_modules/@nx/rspack/src/executors/rspack/rspack.impl.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/config/schema-utils.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/command-line/run/executor-utils.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/devkit-internals.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/utils/assert-workspace-validity.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/project-graph/build-project-graph.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/project-graph/project-graph.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/project-graph/file-utils.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/utils/package-manager.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/utils/package-json.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/utils/print-help.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/command-line/run/run.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/bin/run-executor.js
Pass --verbose to see the stacktrace.

 NX   Running target build for project remote failed

Failed tasks:

- remote:build:production
```

So it looks like it's just missing  '@nx/module-federation/rspack' package, so I triied to install it, but that didn't work. Realiized I need just @nx/module-federation so installed that, and now it works.

So I think the generators need to be updated to install this package.

### Nx Report

```shell
NX   Report complete - copy this into the issue template

Node           : 22.11.0
OS             : darwin-x64
Native Target  : x86_64-macos
pnpm           : 8.10.2

nx                 : 20.2.1
@nx/js             : 20.2.1
@nx/jest           : 20.2.1
@nx/eslint         : 20.2.1
@nx/workspace      : 20.2.1
@nx/devkit         : 20.2.1
@nx/eslint-plugin  : 20.2.1
@nx/react          : 20.2.1
@nx/rspack         : 20.2.1
@nx/web            : 20.2.1
typescript         : 5.6.3
---------------------------------------
Registered Plugins:
@nx/eslint/plugin
@nx/jest/plugin
---------------------------------------
Community plugins:
@nx/module-federation : 20.2.1
```

### Failure Logs

```shell
NX   Cannot find module '@nx/module-federation/rspack'

Require stack:
- /Volumes/Code/react-monorepo/remote/rspack.config.ts
- /Volumes/Code/react-monorepo/remote/rspack.config.prod.ts
- /Volumes/Code/react-monorepo/node_modules/.pnpm/@nx+rspack@20.2.1_@module-federation+enhanced@0.7.6_@module-federation+node@2.6.11_@swc-node+_vvolop54a2iqvroiphed6tvd2q/node_modules/@nx/rspack/src/utils/resolve-user-defined-rspack-config.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/@nx+rspack@20.2.1_@module-federation+enhanced@0.7.6_@module-federation+node@2.6.11_@swc-node+_vvolop54a2iqvroiphed6tvd2q/node_modules/@nx/rspack/src/executors/rspack/lib/config.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/@nx+rspack@20.2.1_@module-federation+enhanced@0.7.6_@module-federation+node@2.6.11_@swc-node+_vvolop54a2iqvroiphed6tvd2q/node_modules/@nx/rspack/src/utils/create-compiler.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/@nx+rspack@20.2.1_@module-federation+enhanced@0.7.6_@module-federation+node@2.6.11_@swc-node+_vvolop54a2iqvroiphed6tvd2q/node_modules/@nx/rspack/src/executors/rspack/rspack.impl.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/config/schema-utils.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/command-line/run/executor-utils.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/devkit-internals.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/utils/assert-workspace-validity.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/project-graph/build-project-graph.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/project-graph/project-graph.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/project-graph/file-utils.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/utils/package-manager.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/utils/package-json.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/utils/print-help.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/src/command-line/run/run.js
- /Volumes/Code/react-monorepo/node_modules/.pnpm/nx@20.2.1_@swc-node+register@1.9.2_@swc+core@1.5.29/node_modules/nx/bin/run-executor.js
Pass --verbose to see the stacktrace.
```

### Package Manager Version

pnpm 8.10.2

### Operating System

- [x] macOS
- [ ] Linux
- [ ] Windows
- [ ] Other (Please specify)

### Additional Information

_No response_

--- a/packages/angular/migrations.json
+++ b/packages/angular/migrations.json
@@ -332,6 +332,12 @@
       },
       "description": "Remove the deprecated 'tailwindConfig' option from ng-packagr executors. Tailwind CSS configurations located at the project or workspace root will be picked up automatically.",
       "factory": "./src/migrations/update-20-2-0/remove-tailwind-config-from-ng-packagr-executors"
+    },
+    "ensure-nx-module-federation-package": {
+      "cli": "nx",
+      "version": "20.3.0-beta.2",
+      "description": "If workspace includes Module Federation projects, ensure the new @nx/module-federation package is installed.",
+      "factory": "./src/migrations/update-20-3-0/ensure-nx-module-federation-package"
     }
   },
   "packageJsonUpdates": {

--- a/packages/angular/src/generators/setup-mf/setup-mf.ts
+++ b/packages/angular/src/generators/setup-mf/setup-mf.ts
@@ -51,6 +51,7 @@ export async function setupMf(tree: Tree, rawOptions: Schema) {
         {
           '@nx/web': nxVersion,
           '@nx/webpack': nxVersion,
+          '@nx/module-federation': nxVersion,
         }
       );
     }
@@ -82,6 +83,7 @@ export async function setupMf(tree: Tree, rawOptions: Schema) {
         {
           '@nx/webpack': nxVersion,
           '@module-federation/enhanced': moduleFederationEnhancedVersion,
+          '@nx/module-federation': nxVersion,
         }
       );
     }

--- a/packages/angular/src/migrations/update-20-2-0/migrate-with-mf-import-to-new-package.ts
+++ b/packages/angular/src/migrations/update-20-2-0/migrate-with-mf-import-to-new-package.ts
@@ -3,9 +3,11 @@ import {
   formatFiles,
   readProjectConfiguration,
   visitNotIgnoredFiles,
+  addDependenciesToPackageJson,
 } from '@nx/devkit';
 import { forEachExecutorOptions } from '@nx/devkit/src/generators/executor-options-utils';
 import { tsquery } from '@phenomnomnominal/tsquery';
+import { nxVersion } from '../../utils/versions';
 
 const NX_ANGULAR_MODULE_FEDERATION_IMPORT_SELECTOR =
   'ImportDeclaration > StringLiteral[value=@nx/angular/module-federation], VariableStatement CallExpression:has(Identifier[name=require]) > StringLiteral[value=@nx/angular/module-federation]';
@@ -51,6 +53,16 @@ export default async function migrateWithMfImport(tree: Tree) {
 
       tree.write(filePath, contents);
     });
+  }
+
+  if (projects.size !== 0) {
+    addDependenciesToPackageJson(
+      tree,
+      {},
+      {
+        '@nx/module-federation': nxVersion,
+      }
+    );
   }
 
   await formatFiles(tree);

--- a/packages/angular/src/migrations/update-20-3-0/ensure-nx-module-federation-package.ts
+++ b/packages/angular/src/migrations/update-20-3-0/ensure-nx-module-federation-package.ts
@@ -1,0 +1,30 @@
+import {
+  type Tree,
+  readProjectConfiguration,
+  addDependenciesToPackageJson,
+} from '@nx/devkit';
+import { forEachExecutorOptions } from '@nx/devkit/src/generators/executor-options-utils';
+import { nxVersion } from '../../utils/versions';
+
+export default async function ensureMfPackage(tree: Tree) {
+  const projects = new Set<string>();
+
+  forEachExecutorOptions(
+    tree,
+    '@nx/angular:module-federation-dev-server',
+    (options, project, target) => {
+      const projectConfig = readProjectConfiguration(tree, project);
+      projects.add(projectConfig.root);
+    }
+  );
+
+  if (projects.size !== 0) {
+    addDependenciesToPackageJson(
+      tree,
+      {},
+      {
+        '@nx/module-federation': nxVersion,
+      }
+    );
+  }
+}

--- a/packages/react/migrations.json
+++ b/packages/react/migrations.json
@@ -41,6 +41,12 @@
       "version": "20.2.0-beta.2",
       "description": "Update the withModuleFederation import use @nx/module-federation/webpack.",
       "factory": "./src/migrations/update-20-2-0/migrate-with-mf-import-to-new-package"
+    },
+    "ensure-nx-module-federation-package": {
+      "cli": "nx",
+      "version": "20.3.0-beta.2",
+      "description": "If workspace includes Module Federation projects, ensure the new @nx/module-federation package is installed.",
+      "factory": "./src/migrations/update-20-3-0/ensure-nx-module-federation-package"
     }
   },
   "packageJsonUpdates": {

--- a/packages/react/src/generators/host/host.ts
+++ b/packages/react/src/generators/host/host.ts
@@ -153,6 +153,7 @@ export async function hostGenerator(
     { '@module-federation/enhanced': moduleFederationEnhancedVersion },
     {
       '@nx/web': nxVersion,
+      '@nx/module-federation': nxVersion,
     }
   );
   tasks.push(installTask);

--- a/packages/react/src/generators/remote/remote.ts
+++ b/packages/react/src/generators/remote/remote.ts
@@ -226,6 +226,7 @@ export async function remoteGenerator(host: Tree, schema: Schema) {
     {
       '@module-federation/enhanced': moduleFederationEnhancedVersion,
       '@nx/web': nxVersion,
+      '@nx/module-federation': nxVersion,
     }
   );
   tasks.push(installTask);

--- a/packages/react/src/migrations/update-20-2-0/migrate-with-mf-import-to-new-package.ts
+++ b/packages/react/src/migrations/update-20-2-0/migrate-with-mf-import-to-new-package.ts
@@ -3,9 +3,11 @@ import {
   formatFiles,
   readProjectConfiguration,
   visitNotIgnoredFiles,
+  addDependenciesToPackageJson,
 } from '@nx/devkit';
 import { forEachExecutorOptions } from '@nx/devkit/src/generators/executor-options-utils';
 import { tsquery } from '@phenomnomnominal/tsquery';
+import { nxVersion } from '../../utils/versions';
 
 const NX_RSPACK_MODULE_FEDERATION_IMPORT_SELECTOR =
   'ImportDeclaration > StringLiteral[value=@nx/react/module-federation], VariableStatement CallExpression:has(Identifier[name=require]) > StringLiteral[value=@nx/react/module-federation]';
@@ -51,6 +53,16 @@ export default async function migrateWithMfImport(tree: Tree) {
 
       tree.write(filePath, contents);
     });
+  }
+
+  if (projects.size !== 0) {
+    addDependenciesToPackageJson(
+      tree,
+      {},
+      {
+        '@nx/module-federation': nxVersion,
+      }
+    );
   }
 
   await formatFiles(tree);

--- a/packages/react/src/migrations/update-20-3-0/ensure-nx-module-federation-package.ts
+++ b/packages/react/src/migrations/update-20-3-0/ensure-nx-module-federation-package.ts
@@ -1,0 +1,30 @@
+import {
+  type Tree,
+  readProjectConfiguration,
+  addDependenciesToPackageJson,
+} from '@nx/devkit';
+import { forEachExecutorOptions } from '@nx/devkit/src/generators/executor-options-utils';
+import { nxVersion } from '../../utils/versions';
+
+export default async function ensureMfPackage(tree: Tree) {
+  const projects = new Set<string>();
+
+  forEachExecutorOptions(
+    tree,
+    '@nx/react:module-federation-dev-server',
+    (options, project, target) => {
+      const projectConfig = readProjectConfiguration(tree, project);
+      projects.add(projectConfig.root);
+    }
+  );
+
+  if (projects.size !== 0) {
+    addDependenciesToPackageJson(
+      tree,
+      {},
+      {
+        '@nx/module-federation': nxVersion,
+      }
+    );
+  }
+}

--- a/packages/rspack/migrations.json
+++ b/packages/rspack/migrations.json
@@ -5,6 +5,12 @@
       "version": "20.2.0-beta.3",
       "description": "Update the withModuleFederation import use @nx/module-federation/rspack.",
       "factory": "./src/migrations/update-20-2-0/migrate-with-mf-import-to-new-package"
+    },
+    "ensure-nx-module-federation-package": {
+      "cli": "nx",
+      "version": "20.3.0-beta.2",
+      "description": "If workspace includes Module Federation projects, ensure the new @nx/module-federation package is installed.",
+      "factory": "./src/migrations/update-20-3-0/ensure-nx-module-federation-package"
     }
   },
   "packageJsonUpdates": {

--- a/packages/rspack/src/migrations/update-20-2-0/migrate-with-mf-import-to-new-package.ts
+++ b/packages/rspack/src/migrations/update-20-2-0/migrate-with-mf-import-to-new-package.ts
@@ -3,9 +3,11 @@ import {
   formatFiles,
   readProjectConfiguration,
   visitNotIgnoredFiles,
+  addDependenciesToPackageJson,
 } from '@nx/devkit';
 import { forEachExecutorOptions } from '@nx/devkit/src/generators/executor-options-utils';
 import { tsquery } from '@phenomnomnominal/tsquery';
+import { nxVersion } from '../../utils/versions';
 
 const NX_RSPACK_MODULE_FEDERATION_IMPORT_SELECTOR =
   'ImportDeclaration > StringLiteral[value=@nx/rspack/module-federation], VariableStatement CallExpression:has(Identifier[name=require]) > StringLiteral[value=@nx/rspack/module-federation]';
@@ -51,6 +53,16 @@ export default async function migrateWithMfImport(tree: Tree) {
 
       tree.write(filePath, contents);
     });
+  }
+
+  if (projects.size !== 0) {
+    addDependenciesToPackageJson(
+      tree,
+      {},
+      {
+        '@nx/module-federation': nxVersion,
+      }
+    );
   }
 
   await formatFiles(tree);

--- a/packages/rspack/src/migrations/update-20-3-0/ensure-nx-module-federation-package.ts
+++ b/packages/rspack/src/migrations/update-20-3-0/ensure-nx-module-federation-package.ts
@@ -1,0 +1,30 @@
+import {
+  type Tree,
+  readProjectConfiguration,
+  addDependenciesToPackageJson,
+} from '@nx/devkit';
+import { forEachExecutorOptions } from '@nx/devkit/src/generators/executor-options-utils';
+import { nxVersion } from '../../utils/versions';
+
+export default async function ensureMfPackage(tree: Tree) {
+  const projects = new Set<string>();
+
+  forEachExecutorOptions(
+    tree,
+    '@nx/rspack:module-federation-dev-server',
+    (options, project, target) => {
+      const projectConfig = readProjectConfiguration(tree, project);
+      projects.add(projectConfig.root);
+    }
+  );
+
+  if (projects.size !== 0) {
+    addDependenciesToPackageJson(
+      tree,
+      {},
+      {
+        '@nx/module-federation': nxVersion,
+      }
+    );
+  }
+}

